File: p031m/next_permutation.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var nextPermutation = function(nums) {
  // 1. Find the ascending neighbors
  let i = nums.length - 1;
  while (i >= 1 && nums[i - 1] >= nums[i]) {
    // >= will handle the duplicated permutation
    i--;
  }
  // 2. Swap
  let j = nums.length - 1;
  if (i > 0) {
    while (nums[i - 1] >= nums[j]) {
      j--;
    }
    [nums[i - 1], nums[j]] = [nums[j], nums[i - 1]];
  }
  // 3. Reverse
  j = nums.length - 1;
  for (; i < j; ++i, --j) {
    [nums[i], nums[j]] = [nums[j], nums[i]];
  }
};
// SPOT TEST
console.log('SPOT TEST:');
[
  [1, 3, 2],
  [1, 2, 3],
  [9, 1, 1],
  [2, 3, 4, 1],
  [5, 4, 3, 2, 1],
  [1, 5, 8, 4, 7, 6, 5, 3, 2, 1]
].forEach(t => {
  console.log('Given', t);
  nextPermutation(t);
  console.log('Next ', t);
  console.log('------');
});
// RANGE TEST
console.log('RANGE TEST:');
[2, 3, 4].forEach(t => {
  let fact = function(n) {
    if (n === 2) return 2;
    return n * fact(n - 1);
  };
  console.log('The next permutations of order', t, '->');
  let perm = Array.from({ length: t }, (_, k) => t - k);
  for (let i = 1; i <= fact(t) + 1; ++i) {
    nextPermutation(perm);
    console.log('  ', perm);
  }
});

File: p031m/next_permutation.py
from typing import List


class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        i = len(nums) - 1
        while i >= 1 and nums[i - 1] >= nums[i]:
            i -= 1
        j = len(nums) - 1
        if i > 0:
            while nums[i - 1] >= nums[j]:
                j -= 1
            nums[i - 1], nums[j] = nums[j], nums[i - 1]
        j = len(nums) - 1
        while i < j:
            nums[i], nums[j] = nums[j], nums[i]
            i += 1
            j -= 1


# TESTS
tests = [
    ([1, 3, 2], [2, 1, 3]),
    ([1, 2, 3], [1, 3, 2]),
    ([9, 1, 1], [1, 1, 9]),
    ([2, 3, 4, 1], [2, 4, 1, 3]),
    ([5, 4, 3, 2, 1], [1, 2, 3, 4, 5]),
    ([1, 5, 8, 4, 7, 6, 5, 3, 2, 1], [1, 5, 8, 5, 1, 2, 3, 4, 6, 7]),
]
for t in tests:
    sol = Solution()
    print("Next permutation of", t[0], "->", t[1])
    sol.nextPermutation(t[0])
    assert t[0] == t[1]
